### PR TITLE
fixed bandwidth selection per month

### PR DESCRIPTION
**Pull request by ArneBab**
I changed the scheme to not rely on the absolute value of the min bandwidth, so it does not break again when we adjust the min bandwidth.

### CODE REVIEW

**Comment by ArneBab** [comment]
@Thynix @nextgens could you have a look?

File: src/freenet/clients/http/wizardsteps/BANDWIDTH_MONTHLY.java [hunk 1]
```diff
@@ -121,24 +121,22 @@ public String postStep(HTTPRequest request)  {
...
 		/*
-		 * Fraction of total limit used for download. Linear from 0.5 at the minimum cap to 0.8 at 100 GB.
-		 *
-		 * This is a line between (minCap, minFraction) and (maxScaleCap, maxFraction) which has been shifted
-		 * to the left by minCap.
+		 * Fraction of total limit used for download. Asymptotically from 0.5 at the minimum cap to 0.8.
+         * 
+         * FIXME: Why do we do this? It does not actually work, since
+         * download cannot be larger than upload for any long amount
+         * of time.
```
Comment (bertm) [inline src/freenet/clients/http/wizardsteps/BANDWIDTH_MONTHLY.java]: Whitespace here is inconsistent

File: src/freenet/clients/http/wizardsteps/BANDWIDTH_MONTHLY.java [hunk 2]
```diff
@@ -121,24 +121,22 @@ public String postStep(HTTPRequest request)  {
...
-		// No minimum fraction clamp: caps less than the minimum get redirected to a warning instead.
-		if (downloadFraction > maxFraction) downloadFraction = maxFraction;
-		double bytesPerSecondTotal = bytesPerMonth/secondsPerMonth;
-		String downloadLimit = String.valueOf(Math.ceil(bytesPerSecondTotal*downloadFraction));
-		String uploadLimit = String.valueOf(Math.ceil(bytesPerSecondTotal*(1-downloadFraction)));
+		double bytesPerSecond = bytesPerMonth/secondsPerMonth;
+        double minBytesPerSecond = Node.getMinimumBandwidth();
+        double bwinc = bytesPerSecond - 2*minBytesPerSecond; // min for up and min for down
+        double dllimit = minBytesPerSecond + (bwinc * ((4./5.)));
+        double ullimit = minBytesPerSecond + (bwinc / 5.);
```
Comment (bertm) [inline src/freenet/clients/http/wizardsteps/BANDWIDTH_MONTHLY.java]: Why not `bwinc * asymptoticFraction` and `bwinc * (1 - asymptoticFraction)`? That avoids having two dependent constants.

Comment (bertm) [inline src/freenet/clients/http/wizardsteps/BANDWIDTH_MONTHLY.java]: Also whitespace

File: src/freenet/clients/http/wizardsteps/BANDWIDTH_MONTHLY.java [hunk 3]
```diff
@@ -121,24 +121,22 @@ public String postStep(HTTPRequest request)  {
...
-		assert minCap < maxScaleCap;
-		double downloadFraction = ((maxFraction-minFraction)/(maxScaleCap - 2*minCap))*(GBPerMonth - minCap) + minFraction;
-		// No minimum fraction clamp: caps less than the minimum get redirected to a warning instead.
-		if (downloadFraction > maxFraction) downloadFraction = maxFraction;
-		double bytesPerSecondTotal = bytesPerMonth/secondsPerMonth;
-		String downloadLimit = String.valueOf(Math.ceil(bytesPerSecondTotal*downloadFraction));
-		String uploadLimit = String.valueOf(Math.ceil(bytesPerSecondTotal*(1-downloadFraction)));
+		double bytesPerSecond = bytesPerMonth/secondsPerMonth;
+        double minBytesPerSecond = Node.getMinimumBandwidth();
+        double bwinc = bytesPerSecond - 2*minBytesPerSecond; // min for up and min for down
```
Comment (bertm) [inline src/freenet/clients/http/wizardsteps/BANDWIDTH_MONTHLY.java]: What if this gets negative? Shoult we report an error?

File: src/freenet/clients/http/wizardsteps/BANDWIDTH_MONTHLY.java (hunk 1 above)
Comment (ArneBab) [inline src/freenet/clients/http/wizardsteps/BANDWIDTH_MONTHLY.java]: forgot the tabify in the comment — thanks! (fixed)

File: src/freenet/clients/http/wizardsteps/BANDWIDTH_MONTHLY.java (hunk 3 above)
Comment (ArneBab) [inline src/freenet/clients/http/wizardsteps/BANDWIDTH_MONTHLY.java]: Bert Massop writes:

>>  		 *
>>  		 * This 50/50 split is consistent with the assumption in the definition of minCap that the upload and
>>  		 * download limits are equal.
>>  		 */
>> -		final double minFraction = 0.5;
>> -		final double maxFraction = 0.8;
>> -		final double maxScaleCap = 100;
>> -		assert minCap < maxScaleCap;
>> -		double downloadFraction = ((maxFraction-minFraction)/(maxScaleCap - 2*minCap))*(GBPerMonth - minCap) + minFraction;
>> -		// No minimum fraction clamp: caps less than the minimum get redirected to a warning instead.
>> -		if (downloadFraction > maxFraction) downloadFraction = maxFraction;
>> -		double bytesPerSecondTotal = bytesPerMonth/secondsPerMonth;
>> -		String downloadLimit = String.valueOf(Math.ceil(bytesPerSecondTotal*downloadFraction));
>> -		String uploadLimit = String.valueOf(Math.ceil(bytesPerSecondTotal*(1-downloadFraction)));
>> +		double bytesPerSecond = bytesPerMonth/secondsPerMonth;
>> +        double minBytesPerSecond = Node.getMinimumBandwidth();
>> +        double bwinc = bytesPerSecond - 2*minBytesPerSecond; // min for up and min for down
>
> What if this gets negative? Shoult we report an error?

If it gets negative the later check to set the bandwidth should fail and
report an error (offering to set the minimum bw).

Best wishes,
Arne
-- 
Unpolitisch sein
heißt politisch sein
ohne es zu merken

File: src/freenet/clients/http/wizardsteps/BANDWIDTH_MONTHLY.java (hunk 2 above)
Comment (ArneBab) [inline src/freenet/clients/http/wizardsteps/BANDWIDTH_MONTHLY.java]: Bert Massop writes:

>>  		 *
>>  		 * This 50/50 split is consistent with the assumption in the definition of minCap that the upload and
>>  		 * download limits are equal.
>>  		 */
>> -		final double minFraction = 0.5;
>> -		final double maxFraction = 0.8;
>> -		final double maxScaleCap = 100;
>> -		assert minCap < maxScaleCap;
>> -		double downloadFraction = ((maxFraction-minFraction)/(maxScaleCap - 2*minCap))*(GBPerMonth - minCap) + minFraction;
>> -		// No minimum fraction clamp: caps less than the minimum get redirected to a warning instead.
>> -		if (downloadFraction > maxFraction) downloadFraction = maxFraction;
>> -		double bytesPerSecondTotal = bytesPerMonth/secondsPerMonth;
>> -		String downloadLimit = String.valueOf(Math.ceil(bytesPerSecondTotal*downloadFraction));
>> -		String uploadLimit = String.valueOf(Math.ceil(bytesPerSecondTotal*(1-downloadFraction)));
>> +		double bytesPerSecond = bytesPerMonth/secondsPerMonth;
>> +        double minBytesPerSecond = Node.getMinimumBandwidth();
>> +        double bwinc = bytesPerSecond - 2*minBytesPerSecond; // min for up and min for down
>> +        double dllimit = minBytesPerSecond + (bwinc * ((4./5.)));
>> +        double ullimit = minBytesPerSecond + (bwinc / 5.);
>
> Also whitespace

Thanks! I’ll fix that.
-- 
Unpolitisch sein
heißt politisch sein
ohne es zu merken

File: src/freenet/clients/http/wizardsteps/BANDWIDTH_MONTHLY.java (hunk 3 above)
Comment (bertm) [inline src/freenet/clients/http/wizardsteps/BANDWIDTH_MONTHLY.java]: Excellent.

File: src/freenet/clients/http/wizardsteps/BANDWIDTH_MONTHLY.java (hunk 2 above)
Comment (ArneBab) [inline src/freenet/clients/http/wizardsteps/BANDWIDTH_MONTHLY.java]: `asymptoticFraction` sounds good: that makes it obvious that the limits always add up to 1.

**Comment by ArneBab** [comment]
There’s something more to this. When I set “I have a fixed bandwidth limit per month [Yes]”, Freenet is extremely slow, even though the speeds are reasonable. I also only get minimal input and output rates, despite high limits:

```
Input Rate: 675 B/s (of 144 KiB/s)
Output Rate: 547 B/s (of 42.9 KiB/s)
```

I can change that by running the wizard again and selecting that I have no monthly limit.

**Comment by ArneBab** [comment]
The slowness after setting monthly limits is gone after a restart of my node. I assume that I ran into some averaging issues since the bandwidth set with a monthly limit was lower than the rate set directly. I need to test this with a new install, but pending that, this pull-request should be merge-ready.

**Comment by ArneBab** [comment]
It also works with a newly installed Freenet. So from my side, this pull request is ready for merging.

**Comment by bertm** [comment]
Looks good to me.

**Comment by ArneBab** [comment]
Thank you!

**Pull request merged**